internal/dcache/cluster_manager/cluster_manager.go

```diff
@@ -961,6 +964,98 @@ func (cmi *ClusterManager) updateStorageClusterMapWithMyRVs() error {
+
+// This method checks if the cluster map update is stuck due to a crashed node, and if so, safely resets the ownership to allow this node to proceed.
+//
+// A cluster map update may get stuck in StateChecking if the node that started the update crashes.
+// This method determines staleness in two ways:
+// 1. Ownership is taken by this node only in another thread.
+// 2. Time-based: The last updated timestamp exceeds a configured threshold.
+// 3. Liveness check: The node that initiated the update is no longer alive (based on heartbeat).
+//
+// There are three possible recovery strategies:
+// 1. Retry the UpdateClusterMapStart using the latest ETag.
+// 2. Forcefully take ownership by setting LastUpdatedBy to this node with existing state checking.
+// 3. Safely reset the cluster map state to StateReady, allowing a clean update to begin.
+//
+// This method uses the third (and most robust) approach: ending the previous update cleanly
+// so the cluster map can be safely updated by this node.
+func (cmi *ClusterManager) overrideClusterMapIfStale(clusterMap *dcache.ClusterMap) (bool, error) {
+
+	// Check if ownership is taken by this node only in another thread.
+	if clusterMap.LastUpdatedBy == cmi.myNodeId {
+		return false, nil
+	}
+
+	//  Check if the last updated timestamp exceeds a configured threshold.
+	staleThreshold := time.Duration(int(cmi.config.ClustermapEpoch)+thresholdClusterMapEpochTime) * time.Second
+	age := time.Since(time.Unix(clusterMap.LastUpdatedAt, 0))
+	if age < staleThreshold {
+		return false, nil
+	}
+
+	log.Warn("ClusterMap last updated at %d (%s ago) exceeds stale threshold %s",
+		clusterMap.LastUpdatedAt, age.Truncate(time.Second), staleThreshold)
+
+	// Optional: check heartbeat of the node that set StateChecking
+	bytes, err := getHeartbeat(clusterMap.LastUpdatedBy)
+	if err != nil {
+		log.Warn("Failed to get heartbeat for node %s: %v", clusterMap.LastUpdatedBy, err)
+		return false, nil
+	}
+
+	var hbData dcache.HeartbeatData
+	if err := json.Unmarshal(bytes, &hbData); err != nil {
+		common.Assert(false, err)
+		log.Warn("Failed to parse heartbeat bytes (%d) for node %s: %v",
+			len(bytes), clusterMap.LastUpdatedBy, err)
+		return false, nil
+
+	}
+
+	// if this node's last heartbeat is too old, consider it dead
+	if !cmi.isHeartBeatStale(hbData) {
+		return false, nil
+	}
+	log.Warn("ClusterManager::overrideClusterMapIfStale: clustermap is stuck in StateChecking by node %s. Overriding Ownership by node %s",
+		clusterMap.LastUpdatedBy, cmi.myNodeId)
+	clusterMap.State = dcache.StateReady
+	clusterMap.LastUpdatedBy = cmi.myNodeId
+	clusterMap.LastUpdatedAt = time.Now().Unix()
+	updatedClusterMapBytes, err := json.Marshal(clusterMap)
+	if err != nil {
+		log.Err("ClusterManager::overrideClusterMapIfStale: Marshal failed for clustermap: %v %+v",
+			err, clusterMap)
+		common.Assert(false, err)
+		return false, err
+	}
+
+	if err = mm.UpdateClusterMapEnd(updatedClusterMapBytes); err != nil {
```

Comment:
since UpdateClusterMapEnd() is not atomic we cannot safely call it, some other node may race with us.
We have to follow the pattern for updating clustermap
1. get cluster map with etag
2. cluster map start with etag
3. if successful, cluster map end